### PR TITLE
Pixhawk 6C external I2C bus hack

### PR DESCRIPTION
**Pull request by julianoes**
**This contains two changes:**

1. This swaps the internal baro and mag back to the bus which is both internal an external but configured as external for this case, in the same way as #20151.

2. Also, it adds a hack for fmu-v6c regarding mag calibration: In order for the mag calibration to treat the mag as an internal one, as well as sensors_status_mag and sensors_status_baro to be correct, we need to insert this override to change these two sensors specifically to internal.

FYI @vincentpoont2 

**Tests:**

Mag calibration:
```
...
MAG 396321 EN: 1, offset: [-0.039 -0.009 -0.025], scale: [0.991 1.069 0.981], Internal
```

sensor status topics:
```
nsh> listener sensors_status_mag

TOPIC: sensors_status_mag
 sensors_status_mag
    timestamp: 233004294 (0.008210 seconds ago)
    device_id_primary: 396321 (Type: 0x06, I2C:4 (0x0C))
    device_ids: [396321, 0, 0, 0]
    inconsistency: [0.0000, nan, nan, nan]
    healthy: [True, False, False, False]
    priority: [50, 0, 0, 0]
    enabled: [True, False, False, False]
    external: [False, False, False, False]

nsh> listener sensors_status_baro

TOPIC: sensors_status_baro
 sensors_status_baro
    timestamp: 241376486 (0.003908 seconds ago)
    device_id_primary: 4028193 (Type: 0x3D, I2C:4 (0x77))
    device_ids: [4028193, 0, 0, 0]
    inconsistency: [0.0000, nan, nan, nan]
    healthy: [True, False, False, False]
    priority: [50, 0, 0, 0]
    enabled: [True, False, False, False]
    external: [False, False, False, False]
```

### CODE REVIEW

**Comment by dagar** [comment]
In terms of something we can live with, rather than a board define slipping into common code maybe we could flip it around so that DeviceExternal() respects something like an override list (or call) that the board support could optionally provide?

**Comment by vincentpoont2** [comment]
If this goes in, then we should we remove this https://github.com/PX4/PX4-Autopilot/pull/20198 ?

**Comment by julianoes** [comment]
@dagar I've updated it, have another look. I believe this is cleaner and doesn't have the board specific ifdef.

**Comment by julianoes** [comment]
@vincentpoont2 yes, this replaces #20198, and  #20151.

**Comment by vincentpoont2** [comment]
This https://github.com/PX4/PX4-Autopilot/pull/20198 just got merged and reverted fmuv6c I2C4 back to "internal" so that the board would start Baro & Mag automatically using Main FW. 

https://github.com/PX4/PX4-Autopilot/blob/1080855f4dedf7ba06a4eb20923ab6c2816b0e32/boards/px4/fmu-v6c/src/i2c.cpp#L39

@julianoes Does this need to move back to "external" with this hack?

**Review by davids5 — DISMISSED** [review]
LGTM

**Comment by julianoes** [comment]
@vincentpoont2 yes, it needs to go back to external.

**Comment by vincentpoont2** [comment]
Everything look good on our test with this PR. @julianoes 

> Board sensors: /etc/init.d/rc.board_sensors
> bmi055_accel #0 on SPI bus 1 rotation 4
> bmi055_gyro #0 on SPI bus 1 rotation 4
> icm42688p #0 on SPI bus 1 rotation 6
> ms5611 #0 on I2C bus 4 (external) address 0x77
> ist8310 #0 on I2C bus 4 (external) address 0xC
> ist8310 #1 on I2C bus 1 (external) address 0xE rotation 10
> ms4525do #0 on I2C bus 4 (external) address 0x28

**Review by davids5 — APPROVED** [review]
Looks good to me

File: boards/px4/fmu-v6c/src/i2c.cpp [hunk 1]
```diff
@@ -36,5 +36,20 @@
...
 };
+
+#ifdef BOARD_OVERRIDE_I2C_TREAT_SENSOR_AS_INTERNAL
+bool i2c_treat_sensor_as_internal(uint32_t device_id)
+{
+	// The internal baro and mag on Pixhawk 6C are on an external
+	// bus. On rev 0, the bus is actually exposed externally, on
+	// rev 1+, it is properly internal, however, still marked as
+	// external for compatibility.
+	if (device_id == 396321 || device_id == 4028193) {
```
Comment (dagar) [inline boards/px4/fmu-v6c/src/i2c.cpp:49]: I would use the device type and other constants so this is a little more readable (and potentially less brittle if something changes).

 - DRV_MAG_DEVTYPE_IST8310
 - particular i2c bus
 - particular i2c address
 - etc

File: boards/px4/fmu-v6c/src/i2c.cpp [hunk 2]
```diff
@@ -36,5 +36,20 @@
 constexpr px4_i2c_bus_t px4_i2c_buses[I2C_BUS_MAX_BUS_ITEMS] = {
 	initI2CBusExternal(1),
 	initI2CBusExternal(2),
-	initI2CBusInternal(4),
+	initI2CBusExternal(4),
 };
+
+#ifdef BOARD_OVERRIDE_I2C_TREAT_SENSOR_AS_INTERNAL
+bool i2c_treat_sensor_as_internal(uint32_t device_id)
```
Comment (dagar) [inline boards/px4/fmu-v6c/src/i2c.cpp:43]: Keep the px4_i2c_ prefix naming convention?

File: boards/px4/fmu-v6c/src/i2c.cpp (hunk 1 above)
Comment (julianoes) [reply boards/px4/fmu-v6c/src/i2c.cpp:49]: Yes, I can try to do that. Let's see if I can solve the puzzle.

File: boards/px4/fmu-v6c/src/i2c.cpp (hunk 2 above)
Comment (julianoes) [reply boards/px4/fmu-v6c/src/i2c.cpp:43]: Ok.

**Comment by julianoes** [comment]
@dagar I've addressed your two comments, fixed up the original commit, rebased on master and force pushed.

**Comment by dagar** [comment]
That looks good, I was just hoping to keep it a little more consistent with px4_i2c_bus_external.

How's this? https://github.com/PX4/PX4-Autopilot/commit/60176fef7508596379b3c225b94e52b927063c18

**Comment by julianoes** [comment]
@dagar your "hack" is much nicer. I will revert my changes, cherry-pick it here, and test it.

**Comment by julianoes** [comment]
Tested, the calibration is correct, and the `sensors_status_mag`/`sensors_status_baro` topics as well.

**Comment by julianoes** [comment]
@davids5 nevermind, it's already fixed: c707997b9e8c8766e64d6b91806117e051647a86.

**Pull request merged**